doc/neps/nep-00XX-array-function-protocol.rst

```diff
@@ -0,0 +1,538 @@
+``__array_function__``.
+
+This protocol is intended to be a catch-all for NumPy functionality that
+is not covered by existing protocols, like reductions (like ``np.sum``)
+or universal functions (like ``np.exp``). The semantics are very similar
+to ``__array_ufunc__``, except the operation is specified by an
+arbitrary callable object rather than a ufunc instance and method.
+
+The interface
+~~~~~~~~~~~~~
+
+We propose the following signature for implementations of
+``__array_function__``:
+
+.. code-block:: python
+
+    def __array_function__(self, func, types, args, kwargs)
+
+-  ``func`` is an arbitrary callable exposed by NumPy's public API,
+   which was called in the form ``func(*args, **kwargs)``.
+-  ``types`` is a list of types for all arguments to the original NumPy
+   function call that will be checked for an ``__array_function__``
+   implementation.
+-  The tuple ``args`` and dict ``**kwargs`` are directly passed on from the
+   original call.
+
+Unlike ``__array_ufunc__``, there are no high-level guarantees about the
+type of ``func``, or about which of ``args`` and ``kwargs`` may contain
+an objects implementing the array API. As a convenience for
+``__array_function__`` implementors of the NumPy API, ``types`` contains
+a list of all types that NumPy checks for potential overloads.
+
+Still be determined: what guarantees can we offer for ``types``? Should
+we promise that types are unique, and appear in the order in which they
+are checked?
+
+Example for a project implementing the NumPy API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most implementations of ``__array_function__`` will start with two
+checks: 1. Is the given function something that we know how to overload?
+2. Are all arguments of a type that we know how to handle?
+
+If these conditions hold, ``__array_function__`` should return
+implementation for ``func(*args, **kwargs)``. Otherwise, it should
+return the sentinel value ``NotImplemented``, indicating that the
+function is not implemented by these types.
+
+.. code:: python
+
+    class MyArray:
+        def __array_function__(self, func, types, args, kwargs):
+            if func not in HANDLED_FUNCTIONS:
+                return NotImplemented
+            if not all(issubclass(t, MyArray) for t in types):
+                return NotImplemented
+            return HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+    HANDLED_FUNCTIONS = {
+        np.concatenate: my_concatenate,
+        np.broadcast_to: my_broadcast_to,
+        np.sum: my_sum,
+        ...
+    }
+
+Necessary changes within the Numpy codebase itself
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This will require two changes within the Numpy codebase:
+
+1. A function to inspect available inputs, look for the
+   ``__array_function__`` attribute on those inputs, and call those
+   methods appropriately until one succeeds.
+
+   This is one additional function of moderate complexity.
```

Comment:
Maybe comment here that speed will be of the essence.

Reply:
In particular, the speed for *regular* arrays should not be impacted much even for small arrays.